### PR TITLE
Framework: Upgrade to webpack@4 

### PR DESCRIPTION
**Pull request by samouri**
**summary**

Webpack 4 was [recently released](https://github.com/webpack/webpack/releases/tag/v4.0.0).  Like any major webpack release, there are some big features that will prove to have an impact on both the developers and users of Calypso.  Below I spell out the expected impact of upgrading:

1. Smaller Bundles:  Webpack 4 [makes tree shaking](https://twitter.com/jdalton/status/893109185264500737?lang=en) significantly more powerful.  It allows for packages to declare themselves as `sideEffect: free` which allows for more aggressive pruning of unused modules.
2. Smarter Chunking: There are now three ways that we create chunks: entry points (sections) `CommonsChunkPlugin`, and `asyncLoad`.  Entry points spell out all of the "sub-applications" within Calypso and contains all of their dependencies whereas `CommonsChunkPlugin` is a tool we use to extract as much of the commonalities we can from our different entry points to that we don't need to separately bundle things like `tinmyce` or `client/components/card` in each entrypoint. In WebPack 4 [they've killed](https://gist.github.com/sokra/1522d586b8e5c0f5072d7565c2bee693) all of the manual configuration required for using `CommonsChunkPlugin` in favor of a new smarter plugin: `SplitChunksPlugin`.  The author of Webpack wrote a whole post about [how it works](https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366) and why it is significantly better than what came before it.  tldr: less configuration, much better optimizations in module duplication across entry-points, and better caching.  We will need to adapt our measuring techniques ([icfy](http://iscalypsofastyet.com/)) to understand the new chunking algorithm.
3. ​Faster Builds:crossed_fingers:: I'm a bit dubious that we will actually see a significant build speed up, but I'm holding out hope! I don't understand how the world has come to this, but the only place I know of to find stats on people's build times is [twitter thread](https://medium.com/webpack/webpack-4-released-today-6cdb994702d4).  There are reports of 20% to 90% speed increased.

**to investigate**
- [x] named modules vs hashed module ids. With thread loader, the module names have gotten quite large
- [x] Does hot reloading still work
- [x] Should we turn off auto splitting in dev? How fast is it without it? (Can't turn it off)

**tools to build**
- [x] spit out code required to load for each section and sizes
- [x] update iscalypsofastyet to understand the new stats object

**metrics to gather before/after**
- [x] build times: cold cache, hot cache, and incremental builds
- [x] total size of all js chunks combined 
- [x] total size of js needed to load homepage 

**to test**
- [x] desktop
- [x] docker
- [x] stability of chunks across builds

### CODE REVIEW

**Comment by blowery** [comment]
Beating on this for a bit

**Comment by samouri** [comment]
@blowery: we still want the ability to add in uglification to development if we want to.
we should be able to remove it from out `package.json` though

File: webpack.config.js [hunk 1]
```diff
@@ -105,6 +105,12 @@ const webpackConfig = {
 		chunkFilename: '[name].[chunkhash].min.js', // ditto
 		devtoolModuleFilenameTemplate: 'app:///[resource-path]',
 	},
+	optimization: {
+		splitChunks: {
```
Comment (samouri) [inline webpack.config.js]: we also want `optimization.splitChunks.name: false`

Comment (blowery) [reply webpack.config.js]: maybe. it looks like things moved back to an id-based chunk id, which won't be stable (likely) across builds.

**Comment by blowery** [comment]
Things left to do here:
- [ ] Turn on `splitChunks: 'all'` to get the entry-point split up
- [ ] Spit out the runtime chunk instead of the old manifest. Maybe inline it if we can.
- [ ] Pick up all the parts of the entry point and current chunk and add script tags for them to the index page. Use the new info in assests to do this.
- [ ] See how stable the chunk hashes are when changing code in various parts of calypso.

File: webpack.config.js (hunk 1 above)
Comment (samouri) [reply webpack.config.js]: if we don't get `[name]` out of the split chunks, then even just a chunk using a new vendor module can rename that vendor chunk.

**Comment by samouri** [comment]
explanations of some of the new options and defaults: https://medium.com/webpack/webpack-4-mode-and-optimization-5423a6bc597a

**Comment by samouri** [comment]
generate stats with `npm run -s env -- node ./bin/generate-stats.js`

**Comment by samouri** [comment]
> add a cacheGroup for tinymce to isolate it into one chunk.

Before we land this we should better understanding why `massivemce` was being duplicated.
Its a major heuristic-smell that makes me worried it could be happening in more places.

**Comment by blowery** [comment]
Marking this as ready for review. Still need to test out desktop and see how that behaves, but it's worth having some other folks look at this.

File: webpack.config.js [hunk 2]
```diff
@@ -99,6 +95,7 @@ const babelLoader = {
 const webpackConfig = {
 	bail: ! isDevelopment,
 	entry: {},
+	mode: bundleEnv,
```
Comment (sirreal) [inline webpack.config.js]: Is this safe? Should we do `bundleEnv === 'development' ? 'development' : 'production'`?

There is a `shared` env that could potentially leak in here:

https://github.com/Automattic/wp-calypso/blob/cd87d90ee7afba0ebc18d301fbc3c0c887a40aca/config/_shared.json#L2

**Review by sirreal — COMMENTED** [review]
Thanks for the work here, I'm really excited to see this and hopeful for improved build times 🤞

File: client/document/index.jsx [hunk 3]
```diff
@@ -162,10 +163,13 @@ class Document extends React.Component {
...
-					<script src={ urls.manifest } />
-					<script src={ urls.vendor } />
-					<script src={ urls[ jsFile ] } />
-					{ chunk && <script src={ urls[ chunk ] } /> }
+					<script
+						dangerouslySetInnerHTML={ {
+							__html: manifest,
+						} }
+					/>
+					{ entrypoint.map( asset => <script src={ asset } /> ) }
```
[the remark this answers is not in the document]
Comment (blowery) [reply client/document/index.jsx]: likely!

**Comment by blowery** [comment]
Getting some nonce violations loading login:
screenshot 2018-03-15 10 59 53

File: package.json [hunk 4]
```diff
@@ -168,7 +168,7 @@
     "npm": "5.6.0"
   },
   "scripts": {
-    "preanalyze-bundles": "cross-env-shell CALYPSO_ENV=production npm run -s build-client -- --profile --json > stats.json",
+    "preanalyze-bundles": "cross-env-shell CALYPSO_ENV=production npm run -s build-client && CALYPSO_ENV=production npm run -s env -- node bin/generate-stats.js",
```
[the remark this answers is not in the document]
Comment (blowery) [reply package.json]: It won't. I think we can get rid of generate-stats altogether and use a stats plugin in the main client build instead.

File: webpack.config.js [hunk 5]
```diff
@@ -107,15 +104,39 @@ const webpackConfig = {
 		chunkFilename: '[name].[chunkhash].min.js', // ditto
 		devtoolModuleFilenameTemplate: 'app:///[resource-path]',
 	},
+	optimization: {
+		splitChunks: {
+			chunks: 'all',
+			name: isDevelopment,
+			cacheGroups: {
+				tinymce: {
```
Comment (DanReyLop) [inline webpack.config.js]: Why is this needed? `tinymce` is used by 2 bundles (`post-editor` and `woocommerce`), so it should be split automatically into a different chunk, right?

File: webpack.config.js [hunk 6]
```diff
@@ -106,15 +103,39 @@ const webpackConfig = {
...
 		rules: [
 			{
+				type: 'javascript/auto',
 				test: /\.jsx?$/,
 				exclude: /node_modules[\/\\](?!notifications-panel)/,
-				loader: [ 'happypack/loader' ],
+				use: _.compact( [
+					{
+						loader: 'thread-loader',
+						options: { workers: 3 },
```
[the remark this answers is not in the document]
Comment (samouri) [reply webpack.config.js]: yep

File: webpack.config.js (hunk 5 above)
Comment (blowery) [reply webpack.config.js]: That's what we expected, but it wasn't happening on its own. Try removing it and then look at analyze-bundles output.

**Comment by blowery** [comment]
> named modules vs hashed module ids. With thread loader, the module names have gotten quite large

This was caused by the babel7 loader, somehow. We moved back to babel6 on the mainline and the issue mostly went away. The module ids are still the full path to the module, not relative to the project root like on webpack3 and we may want to move to hashed module ids as a result, to prevent leaking paths on the build server.

File: server/bundler/assets-writer.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+				errorDetails: true,
+				timings: false,
+				reasons: false,
+			} );
+
+			const statsToOutput = {};
+			statsToOutput.publicPath = stats.publicPath;
+			statsToOutput.manifests = {};
+
+			for ( const name in stats.assetsByChunkName ) {
```
Comment (samouri) [inline server/bundler/assets-writer.js]: we can use `_.find` here

File: client/document/index.jsx [hunk 8]
```diff
@@ -156,18 +157,28 @@ class Document extends React.Component {
 					<script
 						type="text/javascript"
 						nonce={ inlineScriptNonce }
+						key={ '__inline_script' }
```
Comment (sirreal) [inline client/document/index.jsx]: Irrelevant [`key`](https://reactjs.org/docs/lists-and-keys.html#keys)? Can we drop it?

File: client/document/index.jsx [hunk 9]
```diff
@@ -156,18 +157,28 @@ class Document extends React.Component {
...
 						} }
 					/>
 
-					{ i18nLocaleScript && <script src={ i18nLocaleScript } /> }
-					<script src={ urls.manifest } />
-					<script src={ urls.vendor } />
-					<script src={ urls[ jsFile ] } />
-					{ chunk && <script src={ urls[ chunk ] } /> }
-					<script type="text/javascript">window.AppBoot();</script>
+					{ i18nLocaleScript && <script key={ i18nLocaleScript } src={ i18nLocaleScript } /> }
```
Comment (sirreal) [inline client/document/index.jsx]: Irrelevant [`key`](https://reactjs.org/docs/lists-and-keys.html#keys)? Can we drop it?

File: client/document/index.jsx [hunk 10]
```diff
@@ -156,18 +157,28 @@ class Document extends React.Component {
...
 					<script
+						key="__manifest"
```
Comment (sirreal) [inline client/document/index.jsx]: Irrelevant [`key`](https://reactjs.org/docs/lists-and-keys.html#keys)? Can we drop it?

File: client/document/index.jsx [hunk 11]
```diff
@@ -156,18 +157,28 @@ class Document extends React.Component {
...
+						nonce={ inlineScriptNonce }
+						dangerouslySetInnerHTML={ {
+							__html: manifest,
+						} }
+					/>
+					{ entrypoint.map( asset => <script key={ asset } src={ asset } /> ) }
+					{ chunkFiles.map( chunk => <script key={ chunk } src={ chunk } /> ) }
+					<script key="__boot" nonce={ inlineScriptNonce } type="text/javascript">
```
Comment (sirreal) [inline client/document/index.jsx]: Irrelevant [`key`](https://reactjs.org/docs/lists-and-keys.html#keys)? Can we drop it?

File: client/document/index.jsx [hunk 12]
```diff
@@ -156,18 +157,28 @@ class Document extends React.Component {
...
+						window.AppBoot();
+					</script>
+					<script
+						key="__console_warning"
```
Comment (sirreal) [inline client/document/index.jsx]: Irrelevant [`key`](https://reactjs.org/docs/lists-and-keys.html#keys)? Can we drop it?

File: client/document/index.jsx [hunk 13]
```diff
@@ -188,6 +199,7 @@ class Document extends React.Component {
 					{ // Load GA only if enabled in the config.
 					appConfig( 'google_analytics_enabled' ) && (
 						<script
+							key="__ga"
```
Comment (sirreal) [inline client/document/index.jsx]: Irrelevant [`key`](https://reactjs.org/docs/lists-and-keys.html#keys)? Can we drop it?

**Review by sirreal — COMMENTED** [review]
Just a few notes where it looks like we're introducing unnecessary [`key`s](https://reactjs.org/docs/lists-and-keys.html#keys).

**Comment by blowery** [comment]
@sirreal why do you think they're irrelevant?

File: server/bundler/assets-writer.js (hunk 7 above)
Comment (blowery) [reply server/bundler/assets-writer.js]: this is a loop to protect us from a future where we have more than one entrypoint, and therefore, more than one manifest.

**Comment by sirreal** [comment]
@blowery My understanding is that as long as elements have something like a stable position, React does not need a key. It's when rendering things like arrays, particularly with `map`, that React requires a key to uniquely identify an element amongst it's siblings.

The docs aren't great and actually leave me doubting a bit, but I'm pretty sure that these keys can safely be removed ([#](https://reactjs.org/docs/lists-and-keys.html#keys), [#](https://reactjs.org/docs/reconciliation.html#keys)).

Note that removing one of the keys from the `map`s in this file will show an warning on the server console, whereas removing the ones I mentioned does not.

File: bin/generate-stats.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+/** @format */
```
Comment (samouri) [inline bin/generate-stats.js]: I think we can delete this file now

File: server/bundler/assets-writer.js (hunk 7 above)
Comment (samouri) [reply server/bundler/assets-writer.js]: ok. i think thats looking towards a future that might happen, but might never happen.  I'm personally not that very into supporting potential future use-cases just in case [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

even so, we could do something like:
```js
const manifestChunks =  Object.keys( stats.assetsByChunkName )
  .filter( name => name.startsWith( 'manifest' ) )
  .map( name => [ name, stats.assetsByChunkName[ name ].source() ] );

statsToOutput.manifests = fromPairs( manifestChunks );
```

Comment (samouri) [reply server/bundler/assets-writer.js]: Whereas my favorite solution would be to code for the now, and write something like:
```js
statsToOutput.manifests = { manifest: stats.assetsByChunkName[ 'manifest' ].source() }
```

File: server/bundler/assets-writer.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+			}
+
+			statsToOutput.entrypoints = stats.entrypoints;
+
+			function fixupPath( f ) {
+				return path.join( stats.publicPath, f );
+			}
+
+			for ( const entrypoint in statsToOutput.entrypoints ) {
+				// remove the manifest
```
Comment (samouri) [inline server/bundler/assets-writer.js]: we also need to remove it from `stats.chunks[ 'build' ].siblings`

File: server/bundler/assets-writer.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+			for ( const name in stats.assetsByChunkName ) {
+				// make the manifest inlineable
+				if ( String( name ).startsWith( 'manifest' ) ) {
+					statsToOutput.manifests[ name ] = compilation.assets[
+						stats.assetsByChunkName[ name ]
+					].source();
+				}
+			}
+
+			statsToOutput.entrypoints = stats.entrypoints;
```
Comment (samouri) [inline server/bundler/assets-writer.js]: maybe we could set it to the right value the first time with some help from lodash:

```js
statsToOutput.entrypoints = mapValues( stats.entrypoints, entry => ( {
  chunks: reject( entry.assets, 'manifest' ),
  assets: reject( entry.assets, 'manifest' )
} ) );
```

File: server/bundler/assets-writer.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+					.map( fixupPath );
+			}
+			statsToOutput.assetsByChunkName = stats.assetsByChunkName;
+
+			for ( const chunkname in statsToOutput.assetsByChunkName ) {
+				const files = statsToOutput.assetsByChunkName[ chunkname ];
+
+				if ( Array.isArray( files ) ) {
+					statsToOutput.assetsByChunkName[ chunkname ] = files.map( fixupPath );
+				} else {
```
Comment (samouri) [inline server/bundler/assets-writer.js]: why is this sometimes not an array?

**Comment by samouri** [comment]
Awesome work @blowery! Very cool to see `AssetsWriter`, I've never really had to work with a webpack plugin's guts before

File: webpack.config.js [hunk 18]
```diff
@@ -109,13 +110,16 @@ const webpackConfig = {
 		splitChunks: {
 			chunks: 'all',
 			name: isDevelopment,
-			cacheGroups: {
+
+			maxAsyncRequests: 20,
+			maxInitialRequests: 5,
+			/*cacheGroups: {
```
Comment (samouri) [inline webpack.config.js]: you can remove the `cacheGroups` now that you've isolated the cause right?

File: server/pages/index.js [hunk 19]
```diff
@@ -183,7 +211,10 @@ function getDefaultContext( request ) {
 		isDebug,
 		badge: false,
 		lang: config( 'i18n_default_locale_slug' ),
-		jsFile: 'build',
+		entrypoint: getAssets().entrypoints.build.assets.filter(
+			asset => ! asset.startsWith( 'manifest' )
+		),
+		manifest: getAssets().manifests.manifest,
```
Comment (samouri) [inline server/pages/index.js]: this is strange.  while the assets writer supports multiple manifest, this page doesn't.

Does desktop need any changes here?

File: webpack.config.js (hunk 18 above)
Comment (blowery) [reply webpack.config.js]: yeah, but it's a bit fragile. we might want to leave in the tinymce cacheGroup just to force it.

**Comment by samouri** [comment]
There are a couple commented out lines.  
 
- What was `HashedModuleIds` for? Should its line be deleted
- What about `NamedModulesPlugin`

File: webpack.config.js [hunk 20]
```diff
@@ -98,6 +96,8 @@ const babelLoader = {
 const webpackConfig = {
 	bail: ! isDevelopment,
 	entry: {},
```
Comment (samouri) [inline webpack.config.js]: entry is the same for prod and dev, might as well inline it here

File: webpack.config.js [hunk 21]
```diff
@@ -106,15 +106,60 @@ const webpackConfig = {
...
+			} ),
+		],
+	},
 	module: {
 		// avoids this warning:
 		// https://github.com/localForage/localForage/issues/577
 		noParse: /[\/\\]node_modules[\/\\]localforage[\/\\]dist[\/\\]localforage\.js$/,
 		rules: [
 			{
+				type: 'javascript/auto',
```
Comment (samouri) [inline webpack.config.js]: shouldn't need the type here

File: webpack.config.js [hunk 22]
```diff
@@ -187,26 +232,26 @@ const webpackConfig = {
...
-		new webpack.NamedChunksPlugin( chunk => {
-			if ( chunk.name ) {
-				return chunk.name;
-			}
-
-			return chunk.modules.map( m => path.relative( m.context, m.request ) ).join( '_' );
-		} ),
-		new NameAllModulesPlugin(),
-		new AssetsPlugin( {
+		//new NameAllModulesPlugin(),
```
Comment (samouri) [inline webpack.config.js]: nix?

File: server/bundler/assets-writer.js (hunk 15 above)
Comment (blowery) [reply server/bundler/assets-writer.js]: done

File: server/bundler/assets-writer.js (hunk 16 above)
Comment (blowery) [reply server/bundler/assets-writer.js]: done

File: server/bundler/assets-writer.js (hunk 17 above)
Comment (blowery) [reply server/bundler/assets-writer.js]: done

**Comment by samouri** [comment]
`npm run analyze-bundles` is currently giving me:
```
> cross-env-shell CALYPSO_ENV=production EMIT_STATS=true npm run -s build-client

(node:68221) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
TypeError: chunk.startsWith is not a function
    at chunks._.reject.chunk (/Users/jake/Documents/a8c/wp-calypso/server/bundler/assets-writer.js:63:52)
    at /Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:10584:37
    at arrayFilter (/Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:574:11)
    at Function.reject (/Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:9741:14)
    at statsToOutput.entrypoints._.mapValues.entry (/Users/jake/Documents/a8c/wp-calypso/server/bundler/assets-writer.js:63:15)
    at /Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:13392:38
    at /Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:4917:15
    at baseForOwn (/Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:3002:24)
    at Function.mapValues (/Users/jake/Documents/a8c/wp-calypso/node_modules/lodash/lodash.js:13391:7)
    at compiler.hooks.afterEmit.tap.compilation (/Users/jake/Documents/a8c/wp-calypso/server/bundler/assets-writer.js:62:34)
    at _err0 (eval at create (/Users/jake/Documents/a8c/wp-calypso/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:14:1)
    at callback (/Users/jake/Documents/a8c/wp-calypso/node_modules/copy-webpack-plugin/dist/index.js:126:17)
    at /Users/jake/Documents/a8c/wp-calypso/node_modules/copy-webpack-plugin/dist/index.js:220:13
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/Users/jake/Documents/a8c/wp-calypso/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:7:1)
    at AsyncSeriesHook.lazyCompileHook [as _callAsync] (/Users/jake/Documents/a8c/wp-calypso/node_modules/tapable/lib/Hook.js:35:21)
    at asyncLib.forEach.err (/Users/jake/Documents/a8c/wp-calypso/node_modules/webpack/lib/Compiler.js:294:27)
    at done (/Users/jake/Documents/a8c/wp-calypso/node_modules/webpack/node_modules/neo-async/async.js:2809:11)
    at /Users/jake/Documents/a8c/wp-calypso/node_modules/webpack/node_modules/neo-async/async.js:2760:7
    at /Users/jake/Documents/a8c/wp-calypso/node_modules/graceful-fs/graceful-fs.js:43:10
    at FSReqWrap.oncomplete (fs.js:135:15)
npm ERR! code ELIFECYCLE

```

File: client/document/index.jsx [hunk 23]
```diff
@@ -162,12 +163,19 @@ class Document extends React.Component {
...
+							__html: manifest,
+						} }
+					/>
+					{ entrypoint.map( asset => <script key={ asset } src={ asset } /> ) }
+					{ chunkFiles.map( chunk => <script key={ chunk } src={ chunk } /> ) }
+					<script nonce={ inlineScriptNonce } type="text/javascript">
+						window.AppBoot();
+					</script>
+					<script
+						nonce={ inlineScriptNonce }
```
[the remark this answers is not in the document]
Comment (blowery) [reply client/document/index.jsx]: they're for CSP protection. login has it turned on.

[the remark this answers is not in the document]
Comment (blowery) [reply client/document/index.jsx]: basically the server generates a nonce which gets injected into an http header and any inline script tags must also have that nonce. Any later scripts that try to add themselves to the page won't have the nonce, so they won't execute. Only applies to inline scripts.

File: webpack.config.js (hunk 22 above)
Comment (blowery) [reply webpack.config.js]: yeah, breaks the naming that happens via optimize somehow. likely either needs an update or it's not really needed. Need to dig up why we used it in the first place.

File: webpack.config.js (hunk 20 above)
Comment (blowery) [reply webpack.config.js]: is it? I thought entry was a weird array in dev. We overwrite it though, so ... could inline prod.

**Comment by blowery** [comment]
Sorry for breaking analyze-bundles. Chunks are sometimes numbers in prod builds, so they don't have a startsWith method... Obviously I missed that. Think I only tested in dev.

**Comment by samouri** [comment]
I think this is ready for 🚢 .
@blowery?

**Review by blowery — APPROVED** [review]
LGTM
